File: backend/src/main/java/br/ufpr/sept/so2/modules/iam/infrastructure/persistence/UsuarioJpaAdapter.java
package br.ufpr.sept.so2.modules.iam.infrastructure.persistence;

import br.ufpr.sept.so2.modules.iam.application.ports.UsuarioRepository;
import br.ufpr.sept.so2.modules.iam.domain.IdentificadorLogin;
import br.ufpr.sept.so2.modules.iam.domain.Usuario;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.UUID;

@Component
public class UsuarioJpaAdapter implements UsuarioRepository {

    private final UsuarioJpaRepository jpaRepository;

    public UsuarioJpaAdapter(UsuarioJpaRepository jpaRepository) {
        this.jpaRepository = jpaRepository;
    }

    @Override
    public Usuario save(Usuario usuario) {
        UsuarioJpaEntity entity = jpaRepository.findById(usuario.getId())
                .orElseGet(() -> UsuarioJpaEntity.fromDomain(usuario));
        entity.merge(usuario);
        return jpaRepository.save(entity).toDomain();
    }

    @Override
    public Optional<Usuario> findById(UUID id) {
        return jpaRepository.findById(id).map(UsuarioJpaEntity::toDomain);
    }

    @Override
    public Optional<Usuario> findByIdentificador(IdentificadorLogin identificador) {
        if (identificador.getTipo() == IdentificadorLogin.Tipo.GRR) {
            return jpaRepository.findByGrrIgnoreCase(identificador.getValor()).map(UsuarioJpaEntity::toDomain);
        }
        return findByEmail(identificador.getValor());
    }

    @Override
    public Optional<Usuario> findByEmail(String email) {
        return jpaRepository.findByAnyEmail(email).map(UsuarioJpaEntity::toDomain);
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/iam/infrastructure/persistence/UsuarioJpaAdapter.kt
package br.ufpr.sept.so2.modules.iam.infrastructure.persistence

import br.ufpr.sept.so2.modules.iam.application.ports.UsuarioRepository
import br.ufpr.sept.so2.modules.iam.domain.IdentificadorLogin
import br.ufpr.sept.so2.modules.iam.domain.Usuario
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import java.util.Optional
import java.util.UUID

@Component
class UsuarioJpaAdapter(
    private val jpaRepository: UsuarioJpaRepository,
) : UsuarioRepository {
    override fun save(usuario: Usuario): Usuario {
        val entity = jpaRepository.findById(usuario.id)
            .orElseGet { UsuarioJpaEntity.fromDomain(usuario) }
        entity.merge(usuario)
        return jpaRepository.save(entity).toDomain()
    }

    override fun findById(id: UUID): Optional<Usuario> =
        jpaRepository.findById(id).map { it.toDomain() }

    override fun findByIdentificador(identificador: IdentificadorLogin): Optional<Usuario> {
        if (identificador.tipo == IdentificadorLogin.Tipo.GRR) {
            return jpaRepository.findByGrrIgnoreCase(identificador.valor).map { it.toDomain() }
        }
        return findByEmail(identificador.valor)
    }

    override fun findByEmail(email: String): Optional<Usuario> =
        jpaRepository.findByAnyEmail(email).map { it.toDomain() }

    override fun findAtivosByAuthority(authority: String): List<Usuario> =
        jpaRepository.findAtivosByAuthority(authority).map { it.toDomain() }

    override fun search(termo: String?, pageable: Pageable): Page<Usuario> {
        val limpo = termo?.trim()?.takeIf { it.isNotEmpty() }
        return jpaRepository.search(limpo, pageable).map { it.toDomain() }
    }

    override fun existsById(id: UUID): Boolean = jpaRepository.existsById(id)
}
